roachprod: update start-tenant command

herkolategan opened the pull request:
The roachprod `start-tenant` command previously only allowed the tenant and host cluster nodes to be disjoint. This is due to the fact that ports and log directories would collide, should a tenant be started on the same node as the host cluster. Changes on this PR and #106497 allow us to remove that requirement now.

Epic: CRDB-18499
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109841)
<!-- Reviewable:end -->

herkolategan committed:
roachprod: add tenant support to dir logic

Previously tenants were only able to start on disjoint clusters using roachprod,
thus allowing it to use the existing node and log directory methods without any
conflicts. Recently changes were made to start accommodating tenants on the same
VM via port management #106497.

This change updates the log logic to include discriminators for the tenant and
instance in the log directory. There is still logic, regarding log retrieval and
process management, that will need to be updated to make full use of the change.

Epic: CRDB-18499
Release note: None

herkolategan committed:
roachprod: remove start-tenant disjoint requirement

Previously port logic did not allow for having both tenants and the host cluster
running on an overlapping set of VMs, as the ports would conflict and node
directories would overlap.

The changes made in #106497 remove the requirement for tenant clusters to be
disjoint. This change updates start-tenant to allow overlapping clusters now
that it is possible for tenants to run on the same VM.

Epic: CRDB-18499
Release note: None

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Small nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109841#-:-Ndap-M32oOjxG17iQij:b-87rnjq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)

___
*[`pkg/roachprod/install/cockroach.go` line 378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109841#-NdaoJ_l0ivVB4FvCyAV:-NdaoJ_l0ivVB4FvCyAW:b-yvsy84) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee14f264989a792760f585a9e73407d9925e4cbd/pkg/roachprod/install/cockroach.go#L378)):*
> ```Go
> 		return filepath.Join(c.localVMDir(node), "data"+suffix)
> 	}
> 	return fmt.Sprintf("data%s", suffix)
> ```

Nit: this is appending the `suffix` to "data" in two places, and doing it in different ways in each case. Suggestion:

```go
dataDir := fmt.Sprintf("data%s", tenantDirSuffix(tenantName, instance))
if c.IsLocal() {
    	return filepath.Join(c.localVMDir(node), dataDir)
}

return dataDir
```
___
*[`pkg/roachprod/install/cockroach.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109841#-NdaohwlCOA1xrdUlGld:-NdaohwlCOA1xrdUlGle:b-juav5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee14f264989a792760f585a9e73407d9925e4cbd/pkg/roachprod/install/cockroach.go#L388)):*
> ```Go
> 		suffix = fmt.Sprintf("-%s-%d", tenantName, instance)
> 	}
> 	return suffix
> ```

Nit: maybe more clear as:

```go
if tenantName != "" && tenantName != SystemTenantName {
    return fmt.Sprintf("-%s-%d", tenantName, instance)
}

return ""
```


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109841)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/roachprod/install/cockroach.go` line 378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109841#-NdaoJ_l0ivVB4FvCyAV:-NdeCLG4D2BFYoUaf-vU:b6etfin) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee14f264989a792760f585a9e73407d9925e4cbd/pkg/roachprod/install/cockroach.go#L378)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: this is appending the `suffix` to "data" in two places, and doing it in different ways in each case. Suggestion:

```go
dataDir := fmt.Sprintf("data%s", tenantDirSuffix(tenantName, instance))
if c.IsLocal() {
    	return filepath.Join(c.localVMDir(node), dataDir)
}

return dataDir
```
</blockquote></details>

True, good suggestion. I'll amend.
___
*[`pkg/roachprod/install/cockroach.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109841#-NdaohwlCOA1xrdUlGld:-NdeCRslB5bUPiSr2AEk:bqe5jid) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee14f264989a792760f585a9e73407d9925e4cbd/pkg/roachprod/install/cockroach.go#L388)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: maybe more clear as:

```go
if tenantName != "" && tenantName != SystemTenantName {
    return fmt.Sprintf("-%s-%d", tenantName, instance)
}

return ""
```
</blockquote></details>

Hah, this is what happens when you refactor something a few times and then forget to re-optimise :)
Thanks for the suggestion! I'll amend.


<!-- Sent from Reviewable.io -->

herkolategan commented:
TFTR!
bors r=renatolabs
